src/sasctl/tasks.py

```diff
@@ -334,43 +332,58 @@ def register_model(
+
+                # Format input & output variable info as lists of dicts
+                input_vars = [
+                    utils.astore.get_variable_properties(v)
+                    for v in result.InputVariables.itertuples()
                 ]
-                for var in in_var:
-                    if not var.get("role"):
-                        var["role"] = "INPUT"
-                out_var = [
-                    utils.astore.get_variable_properties(var)
-                    for var in result.OutputVariables.itertuples()
+                output_vars = [
+                    utils.astore.get_variable_properties(v)
+                    for v in result.OutputVariables.itertuples()
                 ]
-                for var in out_var:
-                    if not var.get("role"):
-                        var["role"] = "OUTPUT"
+
+                # Set the variable 'role' if it wasn't included (not all astores specify)
+                for v in input_vars:
+                    v.setdefault("role", "INPUT")
+                for v in output_vars:
+                    v.setdefault("role", "OUTPUT")
+
+                # Some astores include the target variable in the 'InputVariable' data frame.  Exclude anything not
+                # marked as INPUT.
+                input_vars = [v for v in input_vars if v["role"] == "INPUT"]
+
                 project = _create_project(
-                    project, model_props, repo_obj, in_var, out_var
+                    project, model_props, repo_obj, input_vars, output_vars
                 )
             else:
                 project = mr.get_project(project)
-            astore = conn.astore.download(rstore=model)
-            params = {
-                "name": name,
-                "projectId": project.id,
-                "type": "ASTORE",
-            }
-            model = mr.post(
-                "/models",
-                files={
-                    "files": ("{}.sasast".format(model.params["name"]), astore["blob"])
-                },
-                data=params,
-            )
+
+            if current_session().version_info() < 4:
+                # Upload the model as a ZIP file if using Viya 3.
+                zipfile = utils.create_package(model, input=input)
+                model = mr.import_model_from_zip(
+                    name, project, zipfile, version=version
+                )
+            else:
+                # If using Viya 4, just upload the raw AStore and Model Manager will handle inspection.
+                astore = cas.astore.download(rstore=model)
+                params = {
+                    "name": name,
+                    "projectId": project.id,
+                    "type": "ASTORE",
+                }
+                model = mr.post(
+                    "/models",
+                    files={"files": (f"{model.params['name']}.sasast", astore["blob"])},
+                    data=params,
+                )
```

Comment:
As discussed, at some point we want to move this call into _services.model_repository correct?

Reply:
Yes, some refactoring is definitely needed.  Wanted to get the testing updates implemented first though.